DOCS-779: Add C++ ML modular resource tutorial (#1501)
viamrobotics/docs: docs/program/extend/modular-resources/examples/add-tflite-module.md

# SMEs: Andrew Morrow
---

Viam provides an example [custom module](extend/modular-resources) written in C++ that extends the [ML model](/services/ml/) service to support audio classification using TensorFlow Lite.

The example files can be found in the [Viam C++ SDK](https://github.com/viamrobotics/viam-cpp-sdk):

- [`example_mlmodelservice_tflite.cpp`](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/examples/modules/example_mlmodelservice_tflite.cpp) - a custom module that provides an example `MLModelService` instance which runs TensorFlow Lite models.
- [`example_audio_classification_client.cpp`](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/examples/mlmodel/example_audio_classification_client.cpp) - an example client instance which generates audio samples and invokes the `example_mlmodelservice_tflite` custom module to classify those samples.

This tutorial walks you through everything necessary to start using these example files with your robot, including building the C++ SDK, procuring the necessary support files, configuring your robot and installing `viam-server`, and running the compiled binary example.

> ## Build the C++ SDK

Author: Right on, I'm now advising for both platforms the following:

```
cmake .. -DVIAMCPPSDK_BUILD_TFLITE_EXAMPLE_MODULE=ON -DVIAMCPPSDK_USE_DYNAMIC_PROTOS=ON -DCMAKE_INSTALL_PREFIX=~/example_workspace/opt -G Ninja
```

I should have this ready for another round EoB today!

> ## Build the C++ SDK

Reviewer: This is done, resolving.

To build the Viam C++ SDK, you will need a macOS or Linux computer.
Follow the instructions below for your platform:

{{< tabs >}}
{{% tab name="macOS" %}}

Follow the [Viam C++ SDK build instructions](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/BUILDING.md) to build the SDK on your macOS computer using the `brew` package manager.

While your specific build steps may differ slightly, your installation should generally resemble the following:

1. Install all listed dependencies to support the Viam C++ SDK:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}

>    brew install abseil cmake boost grpc protobuf xtensor pkg-config ninja buf

Reviewer: I'm a little hesitant about re-iterating C++ SDK build instructions with this level of detail, because things are likely to diverge. How will we keep this list in sync as the C++ SDK requirements evolve, etc.? On the other hand, we want the tutorial to be as turnkey as possible so I understand the desire to provide detailed steps.

Author: Definitely valid. To get this working, I had to do some extra stuff not explicitly covered across BUILDING or README in the SDK repo, so I think there's value in being a bit explicit here.

Reviewer: OK.

   ```

1. Clone the Viam C++ SDK to your computer:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   mkdir -p ~/example_workspace
   cd ~/example_workspace
   git clone git@github.com:viamrobotics/viam-cpp-sdk.git
   ```

> 1. Create a <file>build</file> directory to house the build target:

Reviewer: I'd remove `target`. Just `house the build` is fine.

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   cd viam-cpp-sdk/
   mkdir build
   cd build
   ```

> 1. The Viam C++ SDK requires a specific version of `openssl` in order to build successfully.

Reviewer: This isn't actually true as far as I know. I build the SDK against OpenSSL 3. What is required though is to point the C++ SDK build at the brew OpenSSL since it isn't "linked" by brew due to potential conflicts with the system version (which only sort of exists in a ghostly fashion these days). My `CMake` invocation has this in the environment to hook up to the brew OpenSSL in an architecture and ssl-version-independent fashion:

```
PKG_CONFIG_PATH=`brew --prefix`/opt/openssl/lib/pkgconfig
```

Author: Thank you!! Your version saves us a step! And you're correct regarding `openssl` version support, my mistake misreading BUILDING. Thank you!

   Determine the version of `openssl` you are running on your macOS computer:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   brew list | grep openssl
   ```

   If the command returns `openssl@1.1`, proceed to the next step. If the command returns a different `openssl` version, install version `openssl@1.1` specifically:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   brew install openssl@1.1
   ```

1. Create an environment variable `PKG_CONFIG_PATH` which points to the version of `openssl` we will use for the Viam C++ SDK build:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   export PKG_CONFIG_PATH="/opt/homebrew/opt/openssl@1.1/lib/pkgconfig"
   ```

1. Build the C++ SDK by running the following commands:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}

>    cmake .. -DVIAMCPPSDK_USE_DYNAMIC_PROTOS=ON -DCMAKE_INSTALL_PREFIX=/usr/local -G Ninja

Reviewer: I'm not sure it is a good idea to install to `/usr/local`. That's a "real" location on some systems and requires privs to write to, per the `sudo` used below. Maybe `$HOME/opt` is a better default choice. Maybe some language identifying that an installation prefix is suggested but that by default it will just install to the `install` subdirectory of your build.

Or, even better maybe, have it install to `~/example_workspace/opt` or something, since you have already created that location?

Author: Thank you -- great point. I've converted both the macOS and Linux instructions to use `~/example_workspace/opt` instead. 👍

   ninja all
   sudo ninja install
   ```

   This tutorial passes two flags to the build process to help compile successfully:

   - [`VIAMCPPSDK_USE_DYNAMIC_PROTOS`](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/BUILDING.md#viamcppsdk_use_dynamic_protos) to request that proto generation happen along with the build.
   - [`CMAKE_INSTALL_PREFIX`](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/BUILDING.md#cmake_install_prefix) to install to <file>/usr/local/</file> instead of the default <file>./install</file> location.
     To write to the restricted location <file>/usr/local/</file>, the second `ninja` invocation must also use `sudo`.

{{% /tab %}}
{{% tab name="Linux" %}}

> TODO: Linux instructions here. Currently failing at requiring tflite C headers: I have not yet been able to successfully [build the tflite headers on Linux](https://www.tensorflow.org/lite/guide/build_cmake#build_tensorflow_lite_c_library).

Reviewer: Can you provide some details on what sort of problems you are encountering? Or find some time on my calendar and we can pair up on it.

Author: Sure, I've messaged you in Slack with a log of my attempt. I'll find some time for this week -- thanks!

{{% /tab %}}
{{< /tabs >}}

## Download the `yamnet/classification` model file

This example uses the `yamnet/classification` TensorFlow Lite model for audio classification.

> This is a pre-trained model suitable for this example, but you can also [train your own model](/manage/ml/train-model/).

Reviewer: This is a little misleading because I don't think you actually can currently train your own model that would work with the audio classification client unless you produced something with exactly the same semantics as the yamnet/classification model. The client is pretty tied to the model in this case.

Author: Removed the reference to training our own here, and added to "Next steps" section instead.

The `example_audio_classification_client` binary includes a `--generate` function which determines and creates the necessary robot configuration to support this example.

To generate your robot's configuration using `example_audio_classification_client`:

1. First, determine the full path to the `yamnet/classification` model you just downloaded.
   If you followed the instructions above, this path is: <file>~/example_workspace/lite-model_yamnet_classification_tflite_1.tflite</file>.

1. Next, determine the full path to the `example_mlmodelservice_tflite` modular resource example provided with the Viam C++ SDK.
   If you followed the instructions above, this path is: <file>/usr/local/bin/example_mlmodelservice_tflite</file>.

1. Run the `example_audio_classification_client` binary, proving both paths to the `--generate` function in the following fashion:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   example_audio_classification_client --generate --model-path ~/example_workspace/lite-model_yamnet_classification_tflite_1.tflite --tflite-module-path /usr/local/bin/example_mlmodelservice_tflite > ~/example_workspace/viam-example-mlmodel-config.json
   ```

>    If you did not use the [`CMAKE_INSTALL_PREFIX`](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/BUILDING.md#cmake_install_prefix) option when building the C++ SDK, you may need to provide the full path to the `example_audio_classification_client` binary for this step.

Reviewer: `a different` instead of `the full`?

1. If a labels file was provided, labels are read in as a vector of `string` values and the top 5 scores are associated with their labels.

1. Finally, the client runs 100 rounds of inference using the determined label and score pairs, and returns the results of the rounds, including mean and variance values.

Similarly, the `example_mlmodelservice_tflite` custom module can be found at <file>src/viam/examples/modules/example_mlmodelservice_tflite.cpp</file> and also offers rich comments explaining its features and considerations.

## Troubleshooting and additional documentation

* If you experience issues building the C++ SDK, see [C++ SDK: Limitations, Known Issues, and Troubleshooting](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/BUILDING.md).
* To customize your C++ build process or make adjustments to fit your platform or deployment requirements, see [C++ SDK: Options to Configure or Customize the Build](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/BUILDING.md#options-to-configure-or-customize-the-build)

You can find additional reference material in the [C++ SDK documentation](https://cpp.viam.dev/).

> 

Author: Great idea, absolutely: added.

---
title: "Add a TensorFlow Lite Modular Service"
linkTitle: "Add a TensorFlow Lite Modular Service"
weight: 70
type: "docs"
description: "Add a MLModel modular-resource-based service which uses TensorFlow Lite to classify audio samples."
tags: ["ml", "model training", "services"]
# SMEs: Andrew Morrow
---

> Viam provides an example [modular resource](extend/modular-resources) written in C++ that extends the [ML model](/services/ml/) service to run any TensorFlow Lite model.

Reviewer: I think it is worth adding a sentence here to make extra clear that the module is an example, and what the purpose of the example is. Something like: `The example is intended to demonstrate the principles of operation of implementations of MLModel servers and clients, in order to provide a roadmap for users to implement their own. The components used here are for example purposes only and are not intended for production use.` Feel free to adjust to your liking but that's the content I'm hoping to see included.

The example files can be found in the [Viam C++ SDK](https://github.com/viamrobotics/viam-cpp-sdk):

> - [`example_mlmodelservice_tflite.cpp`](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/examples/modules/example_mlmodelservice_tflite.cpp) - an example module which provides a `MLModelService` modular resource capable of running any TensorFlow Lite model.

Reviewer: I see we say "modular resource" later. Maybe say `an example modular resource which provides a `MLModelService` capable ...`

> - [`example_mlmodelservice_tflite.cpp`](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/examples/modules/example_mlmodelservice_tflite.cpp) - an example module which provides a `MLModelService` modular resource capable of running any TensorFlow Lite model.

Reviewer: @npentrel, ok, thanks. Resolving.

- [`example_audio_classification_client.cpp`](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/examples/mlmodel/example_audio_classification_client.cpp) - an example inference client which generates audio samples and invokes the `example_mlmodelservice_tflite` module to classify those samples using the `yamnet/classification` TensorFlow Lite model.

This tutorial walks you through everything necessary to start using these example files with your robot, including building the C++ SDK, configuring your robot and installing `viam-server`, and generating results with the example inference client program.

> ## Build the C++ SDK

Author: Correct, the build steps build both the SDK and its included examples (including this example module + inference client).

I find value in replicating instructions here. Currently the user must read and comprehend [this](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/README.md) and [this](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/BUILDING.md). Expert users may not need this hand holding, but all the rest of us -- for example: me -- probably benefit from having all the gotchas and caveats and probably-affected-bys in one place, imo. I struggled to get a reliable build going in the current alpha state of the SDK, and I feel that users are likely to experience similar.

Perhaps with beta or rc, the in-repo steps can be enhanced (happy to help!) to standardize and land on ease-of-use wins (like making VIAMCPPSDK_USE_DYNAMIC_PROTOS` the default!), but until then I vote to replicate.

> ## Build the C++ SDK

Reviewer 2: Should this then move to program/sdk and then just be linked from here ?

> ## Build the C++ SDK

Author: Here: https://docs.viam.com/program/apis/ ? Hmm, that seems like more of a avail endpoints reference, rather than a build guide. No other SDK building content there either.

We could create a new section that handles this for all SDKs, but then that would definitely dupe other SDK repo content, as with this one.

I'd just advise waiting for cpp SDK to move out of alpha, to start the convo with cpp team about updating their repo's docs, and to then move this section out to README and BUILDING.

I just think we're too much in-between to standardize either side.

> ## Build the C++ SDK

Reviewer 2: No I was thinking here- and then making the install tab very long for c++  https://docs.viam.com/program/

> ## Build the C++ SDK

Author: Gotcha. I see it now, it could definitely live there.

Would you consider it for a separate PR? I think I'd want to involve the CPP team if we're going to try to make a one-size-fits-all (appropriate for /program landing page) build guide. 

The build steps as presented in this tutorial are not one-size-fits all, and would need to be considered with expanded considerations if so, which would be a non-trivial endeavor.

   ```

1. Verify that the resulting configuration file was created successfully:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   cat ~/example_workspace/viam-example-mlmodel-config.json
   ```

1. Copy the contents of this file.
   Then return to your robot's page on [the Viam app](https://app.viam.com), select the **Config** tab, select **Raw JSON**, and paste the configuration into the text area.

1. Click the **Save config** button at the bottom of the page.
   Now, when you switch back to **Builder** mode, you can see the new configuration settings under the **Services** and **Modules** subtabs.

This generated configuration features the minimum required configuration to support this tutorial: `services` parameters for the [ML model](/services/ml/) service and `modules` parameters for the `example_mlmodelservice_tflite` module.

## Run the inference client

With everything configured and running, you can now run the inference client that connects to `viam-server` and uses the `example_mlmodelservice_tflite` module.

> 1. First, determine your robot address and location secret. To do so, navigate to [the Viam app](https://app.viam.com), select the **Code sample** tab, and toggle **Include secret**.

Reviewer: I don't particularly like the workflow of visiting the code samples because it requires digging the secret out of the middle of the code. You can also get the location secret code from the top-level location page, by opening the "Location secret keys" card. I think that might be a better workflow to document here.

Reviewer: @andf-viam - I think this is the only thing left that I think should be improved in this tutorial.

1. The `example_mlmodelservice_tflite` module returns a map of response tensors as a result.

1. The client validates the result, including its expected type: a vector of `float` values.
   The expected output must be defined under `tensor_name_remappings` in your robot configuration for validation to succeed.
   If you followed the instructions above to [generate your robot configuration](#generate-your-robot-configuration), the value `categories` was pre-populated for you in your generated robot configuration.

1. If a labels file was provided, labels are read in as a vector of `string` values and the top 5 scores are associated with their labels.

1. Finally, the client runs 100 rounds of inference using the determined label and score pairs, and returns the results of the rounds, including mean and variance values.

Similarly, the `example_mlmodelservice_tflite` module can be found at <file>src/viam/examples/modules/example_mlmodelservice_tflite.cpp</file> and also offers rich comments explaining its features and considerations.

## Next steps

> This tutorial explores audio classification using the `yamnet/classification` TensorFlow Lite model, but the `MLModelService` modular resource provided by the `example_mlmodelservice_tflite` module can accept any TensorFlow Lite model so long as it fulfils the [TFLite model constraints](https://www.tensorflow.org/lite/models/build#model_design_constraints).

Reviewer 2: I think the words audio classification should also be in the introduction

---
title: "Add a TensorFlow Lite Modular Service"
linkTitle: "Add a TensorFlow Lite Modular Service"
weight: 70
type: "docs"
description: "Add a MLModel modular-resource-based service which uses TensorFlow Lite to classify audio samples."
tags: ["ml", "model training", "services"]
# SMEs: Andrew Morrow
---

> Viam provides an example [modular resource](extend/modular-resources) written in C++ that extends the [ML model](/services/ml/) service to run any TensorFlow Lite model.

Reviewer 3: When reading this intro, it did not become clear to me until line 22 ("This tutorial walks you through...") how this page relates to the mentioned example. Maybe reorder or front-load the "what" of this page a bit more. One idea:
```suggestion
This tutorial guides you through an example [modular resource](extend/modular-resources) written in C++ that extends the [ML model](/services/ml/) service to run any TensorFlow Lite model.
```

The example includes an inference client program as well, which generates audio samples and uses the modular resource to classify the audio samples based on a pre-trained model.

> This example is intended to demonstrate the design, implementation, and usage of a custom module to serve as a guide to help you write your own.

Reviewer 3: ```suggestion
This example demonstrates the design, implementation, and usage of a custom module to help you write your own.
```
[opt]

> This code is for instructional purposes only, and is not intended for production use.

Reviewer 3: What does this actually mean? Is this a licensing thing, and people should know that they're not allowed to copy the code for production use? Or is it a functionality thing, and maybe the code isn't as robust/efficient/mature as one would want for production use?

Reviewer 3: Cool. I'm thinking this could be made more clear to users here.

You can find the example files in the [Viam C++ SDK](https://github.com/viamrobotics/viam-cpp-sdk):

> - [`example_mlmodelservice_tflite.cpp`](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/examples/modules/example_mlmodelservice_tflite.cpp) - an example module which provides a `MLModelService` modular resource capable of running any TensorFlow Lite model.

Reviewer 3: ```suggestion
- [`example_mlmodelservice_tflite.cpp`](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/examples/modules/example_mlmodelservice_tflite.cpp) - an example module which provides an `MLModelService` modular resource capable of running any TensorFlow Lite model.
```
[nit]

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   mkdir -p ~/example_workspace
   cd ~/example_workspace
   git clone git@github.com:viamrobotics/viam-cpp-sdk.git
   ```

1. Create an <file>opt</file> directory to install the build artifacts to:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   mkdir -p ~/example_workspace/opt
   ```

1. Create a <file>build</file> directory to house the build:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   cd viam-cpp-sdk/
   mkdir build
   cd build
   ```

1. Create an environment variable `PKG_CONFIG_PATH` which points to the version of `openssl` installed on your system:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   export PKG_CONFIG_PATH="`brew --prefix`/opt/openssl/lib/pkgconfig"
   ```

1. Build the C++ SDK by running the following commands:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   cmake .. -DVIAMCPPSDK_BUILD_TFLITE_EXAMPLE_MODULE=ON -DVIAMCPPSDK_USE_DYNAMIC_PROTOS=ON -DCMAKE_INSTALL_PREFIX=~/example_workspace/opt -G Ninja
   ninja all
   ninja install
   ```

>    This tutorial passes three flags to the build process to configure the build:

Reviewer 3: On first read I wondered, does the tutorial pass them? Do the steps above pass them? Do I as a user need to be sure to pass them at some point in the future?
```suggestion
   The command above passes three flags to the build process to configure the build:
```

   sudo apt-get install git cmake build-essential libabsl-dev libboost-all-dev libgrpc++-dev libprotobuf-dev pkg-config ninja-build protobuf-compiler-grpc
   ```

1. If you are not using the `bullseye` container included with the SDK, you may need to install a newer version of `cmake` to build the SDK.
   Run the following to determine the version of `cmake` installed on your system:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   cmake --version
   ```

   If the version returned is `3.25` or later, skip to the next step.
   Otherwise, run the following commands to add the `bullseye-backports` repository and install the version of `cmake` provided there:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   sudo apt-get install software-properties-common
   sudo apt-add-repository 'deb http://deb.debian.org/debian bullseye-backports main'
   sudo apt-get update
   sudo apt-get install -t bullseye-backports cmake
   ```

1. Create an <file>opt</file> directory to install the build artifacts to:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   mkdir -p ~/example_workspace/opt/
   ```

> 1. Create a <file>build</file> directory to house the build:

Reviewer 3: Should people put this build directory into their home directory or into the /example_workspace/opt/ directory or where?

Reviewer 3: Oooh just looked at the Mac instructions and it appears it should go in the sdk folder? So if I followed your instructions literally and never cd'ed anywhere that's where it would go (yay!) but I think being explicit about that would be helpful, since you did just have them create a different directory, and we're not actually supposed to cd into that one.

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   mkdir build
   cd build
   ```

1. Build the C++ SDK by running the following commands:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   cmake .. -DVIAMCPPSDK_BUILD_TFLITE_EXAMPLE_MODULE=ON -DVIAMCPPSDK_USE_DYNAMIC_PROTOS=ON -DCMAKE_INSTALL_PREFIX=~/example_workspace/opt -G Ninja
   ninja all
   ninja install
   ```

>    This tutorial passes three flags to the build process to configure the build:

Reviewer 3: ```suggestion
   The command above passes three flags to the build process to configure the build:
```

## Install `viam-server`

Next, install `viam-server` on your robot, if you have not done so already:

1. Navigate to [the Viam app](https://app.viam.com) in your browser and [add a new robot](/manage/fleet/robots/#add-a-new-robot).

1. Switch to the **Setup** tab, and select your platform from the **Mode** selection at the top.
   If you are installing on a Linux system, additionally select your system's **Architecture**.

1. Follow the steps listed under the **Setup** tab to install `viam-server` on your system.

   {{< alert title="Important" color="note" >}}
   If you are installing `viam-server` within the `bullseye` Docker container provided with the C++ SDK, you will need to run the following command *instead* of the command listed in step 2 **Download and install viam-server** in the Viam app:

   ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
   curl https://storage.googleapis.com/packages.viam.com/apps/viam-server/viam-server-stable-x86_64.AppImage -o viam-server && chmod 755 viam-server && sudo ./viam-server --appimage-extract-and-run -config /etc/viam.json
   ```

>    Once installed within the Docker container, you can later run `viam-server` with the following command, from within the directory you installed it to:

Reviewer 3: ```suggestion
  The command above starts `viam-server`.
   If at some point you stop `viam-server`, you can restart it with the following command, from within the directory you installed it to:
```
[opt] I kinda think "you can later run..." is a bit ambiguous. Folks might wonder when "later" is and why they'd need to run it, and whether they need to run it now.

   Then return to your robot's page on [the Viam app](https://app.viam.com), select the **Config** tab, select **Raw JSON**, and paste the configuration into the text area.

1. Click the **Save config** button at the bottom of the page.
   Now, when you switch back to **Builder** mode, you can see the new configuration settings under the **Services** and **Modules** subtabs.

This generated configuration features the minimum required configuration to support this tutorial: `services` parameters for the [ML model](/services/ml/) service and `modules` parameters for the `example_mlmodelservice_tflite` module.

## Run the inference client

With everything configured and running, you can now run the inference client that connects to `viam-server` and uses the `example_mlmodelservice_tflite` module.

1. First, determine your robot address and location secret. To do so, navigate to [the Viam app](https://app.viam.com), select the **Code sample** tab, and toggle **Include secret**.
   The location secret resembles `abcd1234abcd1234abcd1234abcd1234abcd1234abcd1234` and the robot address resembles `my-robot-main.abcdefg123.viam.cloud`.

   {{%  snippet "secret-share.md" %}}

> 1. Then, run the following to start the inference client, providing the necessary access credentials and the path to the labels file extracted earlier:

Reviewer 3: Aren't we meant to discourage folks from putting their secret and address into their command line history?
